File: rust/year_2019/src/day_06.rs
use hashbrown::HashMap;
use std::iter::{self, successors};

pub fn part_1(input: &str) -> usize {
    let children: HashMap<_, Vec<_>> = input.lines().fold(HashMap::new(), |mut orbits, line| {
        let (a, b) = line.split_once(')').unwrap();
        orbits.entry(a).or_default().push(b);
        orbits
    });

    let mut stack = vec![("COM", 0)];
    iter::from_fn(move || {
        stack.pop().map(|(node, depth)| {
            if let Some(children) = children.get(node) {
                for &child in children {
                    stack.push((child, depth + 1));
                }
            }
            depth
        })
    })
    .sum()
}

pub fn part_2(input: &str) -> usize {
    let parent: HashMap<_, _> = input
        .lines()
        .map(|line| {
            let (a, b) = line.split_once(')').unwrap();
            (b, a)
        })
        .collect();

    let ancestors = |start: &str| successors(parent.get(start), |&n| parent.get(n));

    let path_1: HashMap<_, _> = ancestors("YOU")
        .enumerate()
        .map(|(depth, ancestor)| (ancestor, depth))
        .collect();

    ancestors("SAN")
        .enumerate()
        .find_map(|(depth_san, ancestor)| {
            path_1.get(ancestor).map(|&depth_you| depth_you + depth_san)
        })
        .unwrap()
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const EXAMPLE_1: &str = "\
COM)B
B)C
C)D
D)E
E)F
B)G
G)H
D)I
E)J
J)K
K)L";

    const INPUT: &str = include_str!("../test_data/day_06.txt");

    #[test_case(EXAMPLE_1, 42)]
    #[test_case(INPUT, 186597)]
    fn part_1(input: &str, expected: usize) {
        assert_eq!(super::part_1(input), expected);
    }

    const EXAMPLE_2: &str = "\
COM)B
B)C
C)D
D)E
E)F
B)G
G)H
D)I
E)J
J)K
K)L
K)YOU
I)SAN";

    #[test_case(EXAMPLE_2, 4)]
    #[test_case(INPUT, 412)]
    fn part_2(input: &str, expected: usize) {
        assert_eq!(super::part_2(input), expected);
    }
}

File: python/year_2019/day_06.py
"""
https://adventofcode.com/2019/day/6
"""

import re
from collections import defaultdict
from functools import partial
from itertools import takewhile
from operator import ne
from pathlib import Path

from iteration_utilities import applyfunc

R = re.compile(r"(\w+)\)(\w+)")


def part_1(text, com="COM"):
    children = defaultdict(list)
    for line in text.splitlines():
        a, b = R.match(line).groups()
        children[a].append(b)

    stack = [(com, 0)]
    total = 0
    while stack:
        parent, level = stack.pop()
        total += level
        stack.extend((child, level + 1) for child in children[parent])

    return total


def part_2(text, com="COM", you="YOU", san="SAN"):
    parents = {}
    for line in text.splitlines():
        a, b = R.match(line).groups()
        parents[b] = a

    path = takewhile(partial(ne, com), applyfunc(parents.get, you))
    visited = {you: i for i, you in enumerate(path)}

    path = applyfunc(parents.get, san)
    count, dest = next((i, san) for i, san in enumerate(path) if san in visited)

    return visited[dest] + count


def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_06.txt").read_text()


EXAMPLE_1 = """COM)B
B)C
C)D
D)E
E)F
B)G
G)H
D)I
E)J
J)K
K)L
"""

EXAMPLE_2 = """COM)B
B)C
C)D
D)E
E)F
B)G
G)H
D)I
E)J
J)K
K)L
K)YOU
I)SAN
"""


def test_part_1_input():
    assert part_1(_read_input()) == 186597


def test_part_1_example():
    assert part_1(EXAMPLE_1) == 42


def test_part_2_input():
    assert part_2(_read_input()) == 412


def test_part_2_example():
    assert part_2(EXAMPLE_2) == 4
